### ISSUE TITLE
Update logins API to maintain validation/dupe handling

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/application-services#4091]
Update the logins API in order to maintain support for the validation/dupe handling for per-field encryption. As specified in [ADR-0001](https://github.com/mozilla/application-services/blob/main/docs/adr/0001-update-logins-api.md), this work will include the following:

- Replacing login `add` and `update` functions with a new `add_or_update` function. The following should be true of this new function:
  - Will take the encryption key as a parameter
  - Needs to support accurate population of the `time_password_changed` field for login updates
  - Should call `fixup_and_check_for_dupes` as both `add` and `update` currently do
  - Should replicate a call to `potential_dupes_ignoring_username` to distinguish add and update login operations [as occurs in a-c](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L37)
- Removing `check_valid_with_no_dupes` and `potential_dupes_ignoring_username` functions
  - The logic in these functions should be moved to the new `add_or_update` function
- Adding an encryption key parameter to `import_multiple` calls 
- Fixing all broken tests and adding tests where sensible

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2290)
┆Epic: Logins update and sqlcipher migration
┆Fix Versions: Release 93
┆Sprint End Date: 2021-09-03

**Comment by data-sync-user** [comment]
➤ Lougenia Bailey commented:

Status update: Momentarily on hold for the uniffi ticket (UNIFFI-46). Once the uniffi ticket is complete, I’ll start implementing the logins api changes proposed in api shape I’ve solicited feedback for.
